Feature: nordigen bank integration

paulwer opened the pull request:
hey @turbo124,
this is just a first draft for the upcomming implementation and I would like to use this place to ask some questions.

1. Can you hint me to a related resource about enums in laravel. I did not find something matching and had implement "just a string" right now.
2. Right now there is a clientId and clientSecret defined as config variables by me within the model. I think we should move this to the account model, so each company can setup their own account independenly? What do you think about this?
3. Should I create an independent Job for yodlee and nordigen or should the existing job handle both integrations?
4. You guys are doing the front-end implemantations afterwards? ok for you?
5. how do I get the frontend-url (for redirecting-url after registration) or should the frontend provide this as a param?

paulwer committed:
adding webhooks for products

paulwer committed:
first draft

turbo124 commented:
@paulwer 

1. I would recommend declaring class constants as sufficient. This would follow the same pattern that we use elsewhere.
2. Yes, please create a migration on the accounts table, we can use the column name nordigen_keys, you can store the keys as an encrypted keyvalue object 
3. Yes, you'll want to create a separate job for Nordigen, as the Yodlee job runs only on the hosted platform.
4. Yes, when you have completed, let us know and we'll perform the front end work.

The front end will have a CTA to trigger the OAuth flow this involves:

1. Front end calls `/api/v1/one_time_token` POST, with a request body parameter of `context` in this case you'll pass `nordigen` The API will respond with a uuid Hash, you'll then pass this hash into your route which is handling the OAuth flow.
ie. ninja.test/noriden_oauth/:hash

You can then use

```php
       Cache::get($this->hash);
``` 

To resolve your context array which will be in the form of

```
'user_id' => user.id,
'company_key'=> company.company_key,
'context' => 'nordigen'
'is_react' => true | false
```

This hands you back control and you can continue your flow.

You can use the is_react flag to determine which UI to forward the user back to after OAuth, the config helper

config('ninja.react_url') //react
config('ninja.app_url') //flutter

Let us know if you have any questions.

paulwer committed:
fix for seeders

paulwer committed:
wip: changes

paulwer committed:
adding controler method for institutions (wip)

paulwer committed:
feat: institutions endpoint ready
fix: wrong variable names
fix: missing implementation in api-router

paulwer committed:
feat: get AccountData

paulwer committed:
feat: refresh & getAccounts

paulwer commented:
@turbo124 some updates:
1. I successfully fetched an account with balances, created the AccountMapper and created a bank_integration accordingly.
![Unbenannt](https://github.com/invoiceninja/invoiceninja/assets/52678724/2e09013a-c137-47b6-9ad4-44b8dd0f7fbc)
API is also capeable of fetching available institutions of nordigen, to display them for selection in your frontend. (ID must be provided to create a new requisition/auth-process. currently: /api/v1/nordigen/connect)
![Unbenannt](https://github.com/invoiceninja/invoiceninja/assets/52678724/2d7cf6e9-349b-42e2-9c61-a1e277200900)
2. Question/Reinsurance: Nordigen provides multiple balance types:
- closingBooked
- interimAvailable
I am gonna use the first one, because it makes more sense for me to display this value to the user.
3.  I will now create a table called bank_integration_nordigen_requisitions, for storing requisitionIds and determing, which requisition belongs to which account and company. I will not storing the accountIds and instead fetch them from active requisitions of nordigen already. (1. fetch requisitions 2. gattering accountIds 3. fetch accountData)
If a accounts gets unavailable the Job for fetching the transactions will mark the bank_integration as inactive, like already used with yodlee.
4. Question: which bank_account_type should I use for bank accounts from nordigen? yodlee uses 'creditCard', which may not match our needs? Is this a free-text-field. Should ther maybe go the product of the bank_account like: Girokonto Comfort Smart.
5. Help Wanted: Can you take a look at the NordigenControler. When I am using CreateNortigenRequisitionRequest on line 340 (public function connect(CreateNortigenRequisitionRequest $request) I am getting an error, that this class not exists, even, when its imported and named correctly from my POV.
6. Requisitions will expire by default after 90 days. After this period, the frontend should prompt the user to renewal their requisition for their account. With a custom user-aggreement (can be created by the backend, without user-confirmation and just has to be provided within the creation of the requisition) we can take to up to 180 days in max.
7. Question/Reinsurance: I thought about including cleanup functions to get rid of expired requisitions, which can add up easily. If you dont want to include this, it would still work fine and I dont know, if nordigen may deletes them on their own. Should we think about including them?
8. Question: After I developed this feature I also want to use it in the self_hosted version. As of the user is able to provide their own nordigen account, I guess there are no plans to move this under a payed plan in the future?! 😊

turbo124 commented:
@paulwer looks great! Are you able to ping me on slack? I'd like to have a chat about the implementation now that we are diving into details.

regarding self host, this code will be in the public repo and available to all self hosted users.

turbo124 commented on app/Helpers/Bank/Nordigen/Transformer/AccountTransformer.php:
```diff
@@ -0,0 +1,121 @@
+        }
+
+        return [
+            'id' => $nordigen_account->metadata["id"], // TODO: maybe add prefix for unique id between yodlee and nordigen?
```
No need as the integration_type can be used to differentiate

paulwer replied:
I dont know, how Yodlee forms the id. I thought it was the internal resource-id of the provider. Maybe if yodlee and nordigen is used in parallel, there can be overlaps?! Idk.

Because this is the PRIMARY key of the database, there might be the very rare case, we have an overlap?
Thats, why I thought about this.

turbo124 replied:
I think this is safe to leave as is, because we always have the integration type to differentiate, and the column is not unique.

turbo124 commented on app/Helpers/Bank/Nordigen/Transformer/AccountTransformer.php:
```diff
@@ -0,0 +1,121 @@
+
+        return [
+            'id' => $nordigen_account->metadata["id"], // TODO: maybe add prefix for unique id between yodlee and nordigen?
+            'account_type' => "bank_account", // TODO: not creditCard, which type should be used here?!
```
Does Nordigen only return bank_account? if so that is fine, we'll just need to document for the front ends

paulwer replied:
yes, they do.
are there some already used keywords, to set by default from me or should it may be empty?

turbo124 replied:
@paulwer if bank_account is the only keyword used, then this is fine to leave as bank_account.

paulwer replied:
great

turbo124 commented on app/Helpers/Bank/Nordigen/Transformer/AccountTransformer.php:
```diff
@@ -0,0 +1,121 @@
+            'provider_name' => $nordigen_account->institution["name"],
+            'nickname' => $nordigen_account->data["ownerName"] ? $nordigen_account->data["ownerName"] : '',
+            'current_balance' => (int) $used_balance ? $used_balance["balanceAmount"]["amount"] : 0,
+            'account_currency' => $used_balance ? $used_balance["balanceAmount"]["currency"] : '',
```
just confirming this iso 3char currency code

paulwer replied:
yes, you are right. the response is like from the comment above (their api-reference does not show any hint about this):
[balances] => [
                {
                    [balanceAmount]: {
                        [amount] => 9825.64
                        [currency] => EUR
                    },
                    [balanceType] => closingBooked
                    [referenceDate] => 2023-12-01
                },
                {
                    [balanceAmount[: {
                        [amount] => 10325.64
                        [currency] => EUR
                    },
                    [balanceType] => interimAvailable
                    [creditLimitIncluded]: true,
                    [referenceDate] => 2023-12-01
                }
            ]

turbo124 commented on app/Helpers/Bank/Nordigen/Transformer/IncomeTransformer.php:
```diff
@@ -0,0 +1,186 @@
+    {
+        //CREDIT / DEBIT
+
+        if (property_exists($transaction, 'highLevelCategoryId') && $transaction->highLevelCategoryId == 10000012)
```
this looks like a carry over from Yodlee, we'll need a filter that matches Nordigen

paulwer replied:
I will touch this, when I will do the getTransactions() part.
I will close this until then.

turbo124 commented on app/Http/Controllers/Bank/YodleeController.php:
```diff
@@ -14,7 +14,7 @@
 use App\Helpers\Bank\Yodlee\Yodlee;
 use App\Http\Controllers\BaseController;
 use App\Http\Requests\Yodlee\YodleeAuthRequest;
-use App\Jobs\Bank\ProcessBankTransactions;
+use App\Jobs\Bank\ProcessBankTransactionsYodlee;
 use App\Models\BankIntegration;
 use Illuminate\Http\Request;
 
```
Instead of renaming both props, we should only add prefixes for the new providers, so in this case lets leave everything without the yodlee_ prefix, but continue to prefix as nordigen

paulwer replied:
I also touched the prop bank_integration_account_id for yodlee:
$table->renameColumn('bank_integration_account_id', 'bank_integration_yodlee_account_id');

should this also be reverted?

My Idea behind this was to structure both integrations independently, so there is no confusion, which parameter belongs to which one. Exspecialy with future dev.

turbo124 replied:
understood, there are some legacy and platform issues so mutating existing columns needs to be avoided.

paulwer replied:
therefore I will revert the changes in a future commit?

paulwer replied:
i reverted the column at the account table. do your want change of the Job Names?

"understood" could mean both to me🙈

turbo124 replied:
We should leave the yodlee props unchanged, only generate migratoins for the nordigen implementation

paulwer replied:
@turbo124 do the jobnames also be reverted ((of yodlee)?

turbo124 commented on app/Http/Controllers/BankIntegrationController.php:
```diff
@@ -521,23 +524,45 @@ public function bulk(BulkBankIntegrationRequest $request)
-        // if we add additional providers we can reuse this class
+        $account = auth()->user()->account;
+
+        $this->refreshAccountsYodlee($account);
```
due to the return response in this method, code beneath it may never run, i'd suggest just returning if no account_id is present

turbo124 commented on database/factories/BankIntegrationFactory.php:
```diff
@@ -25,6 +26,7 @@ class BankIntegrationFactory extends Factory
     public function definition()
     {
         return [
+            'integration_type' => BankIntegration::INTEGRATION_TYPE_NONE,
```
I think we can drop the none type completely, as we'll set the column to nullable in the migration.

turbo124 commented on routes/api.php:
```diff
@@ -370,9 +371,15 @@
 
-Route::fallback([BaseController::class, 'notFound']);
\ No newline at end of file
+Route::get('api/v1/nordigen/institutions', [NordigenController::class, 'institutions'])->middleware('throttle:100,1')->middleware('token_auth')->name('nordigen_institutions');
+Route::any('api/v1/nordigen/refresh', [NordigenController::class, 'refresh'])->middleware('throttle:100,1')->middleware('token_auth')->name('nordigen_refresh');
+Route::post('api/v1/nordigen/connect', [NordigenController::class, 'connect'])->middleware('throttle:100,1')->middleware('token_auth')->name('nordigen_connect');
+Route::any('api/v1/nordigen/callback', [NordigenController::class, 'callback'])->middleware('throttle:100,1')->name('nordigen_callback');
+
+Route::fallback([BaseController::class, 'notFound']);
```
When out of dev, we'll want to push this up into the closure above

paulwer replied:
done

turbo124 reviewed (commented):
@paulwer 

I've left some comments inline.

paulwer committed:
using nullable as default value for integration_type

paulwer commented:
in the meantime, you should also take a look at a predefined front-end component, which can be a great start for you:
https://developer.gocardless.com/bank-account-data/bank-selection-ui
https://github.com/nordigen/nordigen-bank-ui

paulwer committed:
revert new table, some changes for endpoints, fixes

turbo124 commented:
@paulwer in regards to account_type, can you use `bank` please so that it is consistent with the Yodlee enum for bank account.

paulwer committed:
change account_type

paulwer committed:
feat: connect account

paulwer committed:
renaming Request Dtos

paulwer committed:
renaming again

paulwer committed:
minor fixes

paulwer committed:
rewrite connect/confirm for complete flow usage with only redirects

paulwer committed:
introduce nordigen bank_selection ui

paulwer commented on app/Helpers/Bank/Nordigen/Nordigen.php:
```diff
@@ -0,0 +1,123 @@
+
+    }
+
+    public function getTransactions(string $accountId, string $dateFrom = null)
```
@turbo124 should we include pending transactions within the integration-process and mark them with a specific category?!

paulwer replied:
i just sync booked transactions to keep it simple.

paulwer committed:
latest changes

paulwer commented on resources/views/bank/nordigen/connect.blade.php:
```diff
@@ -0,0 +1,67 @@
+
+    // Pass your redirect link after user has been authorized in institution
+    const config = {
+        // Text that will be displayed on the left side under the logo. Text is limited to 100 characters, and rest will be truncated.
```
@turbo124 please provide text or language key to use as default text here

paulwer replied:
I wrote a default text, which you can use to translate.

paulwer commented on resources/views/bank/nordigen/connect.blade.php:
```diff
@@ -0,0 +1,67 @@
+        styles: {
+            // Primary
+            // Link to google font
+            fontFamily: 'https://fonts.googleapis.com/css2?family=Roboto&display=swap', // @todo replace to match german law: not use google fonts and use local instead
```
@turbo124 does in uses google fonts? are there local hosted versions of it?

turbo124 replied:
/public/fonts/

we do store roboto there, I believe some selfhost users much prefer to load the font locally rather than ping a cdn

paulwer replied:
i replaced it with the following: /assets/fonts/Roboto-Regular.ttf

paulwer committed:
minor additions for IncomeTransformer

paulwer committed:
Merge branch 'v5-develop' of https://github.com/invoiceninja/invoiceninja into v5-develop

paulwer committed:
Merge branch 'v5-develop' into feature-nordigen-payment-provider

paulwer committed:
changes related to merge

paulwer committed:
changes to job-schedule

paulwer commented on app/Jobs/Ninja/BankTransactionSync.php:
```diff
@@ -47,17 +49,29 @@ public function handle()
             MultiDB::setDB($db);
 
-            nlog("syncing transactions");
+            if (Ninja::isSelfHost()) { // @turbo124 @todo I migrated the schedule for the job within the kernel to execute on all platforms and use the same expression here to determine if yodlee can run or not. Please chek/verify
```
@turbo124 @todo I migrated the schedule for the job within the kernel to execute on all platforms and use the same expression here to determine if yodlee can run or not. Please chek/verify

turbo124 replied:
This looks fine

paulwer committed:
fixes

turbo124 commented on app/Jobs/Ninja/BankTransactionSync.php:
```diff
@@ -49,7 +49,7 @@ public function handle()
             MultiDB::setDB($db);
 
-            if (Ninja::isSelfHost()) { // @turbo124 @todo I migrated the schedule for the job within the kernel to execute on all platforms and use the same expression here to determine if yodlee can run or not. Please chek/verify
+            if (Ninja::isHosted()) { // @turbo124 @todo I migrated the schedule for the job within the kernel to execute on all platforms and use the same expression here to determine if yodlee can run or not. Please chek/verify
```
This looks fine, :+1:

paulwer committed:
replace font with local one

paulwer committed:
dev-workaround for storing accountId + updates to transaction jobs + local fonts

paulwer committed:
dedicated keys for nordigen

paulwer committed:
fixes

paulwer committed:
introduce ENV-variables for nordigen

paulwer commented:
@turbo124  ENV-configuration was implemented as follows:
1. check for configuration at account-level
2. check for configuration at ENV level
3. if none exists service is disabled

I also introduced the test-mode property, where you are only able to select the test-bank of nordigen.
This can be used in test-cases or for you to test in local. nordigen will then provide test-bank accounts and test transactions.

paulwer committed:
fixes

paulwer committed:
fixes

paulwer committed:
fix: missing context error

paulwer committed:
fixes

paulwer committed:
fixes

paulwer committed:
fixes

paulwer committed:
fixes

turbo commented:
Wrong person 😉

paulwer commented:
I am sorry :)

paulwer committed:
remove credentials on account level + remove ApiException + fixes

paulwer committed:
fix spelling

paulwer committed:
fixes for env

paulwer committed:
fixes for sync transactions

paulwer committed:
adding additional fields to bank_transactions
https://github.com/invoiceninja/invoiceninja/issues/8042

paulwer committed:
revert yodlee-change: https://github.com/paulwer/invoiceninja/pull/1

paulwer committed:
fixes

paulwer committed:
fixes

paulwer committed:
Merge branch 'v5-develop' of https://github.com/invoiceninja/invoiceninja into feature-nordigen-payment-provider

paulwer commented:
## Guide for setting up your environment for review:
1. Get yourself an account on gocardless/nordigen: https://bankaccountdata.gocardless.com/
2. Create new credentials: https://bankaccountdata.gocardless.com/user-secrets/
3. add them to .env => NORDIGEN_SECRET_ID & NORDIGEN_SECRET_KEY

## Suggestions/Questions/Open Points

### Test Auth-Flow: (without institution selected)
1. create one_time_token
2. open /nordigen/connect/{one_time_token}
3. perform a full flow, f.ex. with the bank: Sandbox Finance
### Test Auth-Flow: (with institution selected, f.ex. for rerequesting auth on disabled bank_integrations)
same like above, but provide institution_id from registered bank_integration from section above
### Check Account-Data & Transactions
the current frontend only displays the description: is there any way to introduce line-breaks?
is CREDIT/DEBIT assigned correctly?
### Check Changed API-Responses for brank_integrations & bank_transactions
- bank_integrations: 
  - i leave nordigen_account_id as not public, for data-security and made nordigen_institution_id public for better display in frontend (f.ex. fetching logo from api/v1/nordigen/institutions or rerequest auth for invalid bank_integration)
  - property disabled_upstream: was already public, so the frontend can determine, which accounts need a new requisition flow performed by the user. (Auth-Flow with institutioId from the bank_integration)
- bank_transactions:
  - debitor & debitor_name: new properties according to #8042 
### Check @todo and @turbo124 comments
there are also some comments left to answer or with todo for you
### Check User Payment
I integrated the jobs/api endpoints with a relation to if the user has access to the services of nordigen. please check this implementation.
### Translations (DONE)
Add Translations for connect.blade.php as mentioned.
### OpenApi Docs
todo

turbo124 reviewed (commented):
Minor notes, only

paulwer committed:
display occured errors in flow with a view

paulwer committed:
translations

paulwer committed:
fixes

paulwer committed:
fixes

paulwer committed:
change debitor to participant

paulwer committed:
remove unnecessary code and change sync-behavior to suppress sync error based on wrong from_date

paulwer committed:
fix composer build

paulwer committed:
update deps, fixes & stop processing archived bank_integrations

paulwer committed:
better logging for jobs

paulwer commented on app/Jobs/Bank/ProcessBankTransactionsYodlee.php:
```diff
@@ -151,14 +157,14 @@ private function processTransactions()
+
         foreach ($transactions as $transaction) {
-            if (BankTransaction::query()->where('transaction_id', $transaction['transaction_id'])->where('company_id', $this->company->id)->withTrashed()->exists()) {
+            if (BankTransaction::query()->where('transaction_id', $transaction['transaction_id'])->where('company_id', $this->company->id)->where('bank_integration_id', $this->bank_integration->id)->withTrashed()->exists()) { // @turbo124 was not scoped to bank_integration_id => from my pov this should be present, because when an account was historized (is_deleted) a transaction can occur multipl ...
```
@turbo124 was not scoped to bank_integration_id => from my pov this should be present, because when an account was historized (is_deleted) a transaction can occur multiple (in the archived bank_integration and in the new one

turbo124 replied:
I think this links back to whether we create new integration, or restore existing. This will determine which way to handle this.

paulwer committed:
some changes related to logging & cleanup

paulwer commented on app/Http/Controllers/BankIntegrationController.php:
```diff
@@ -288,12 +336,20 @@ public function removeAccount(AdminBankIntegrationRequest $request, $acc_id)
-            (new ProcessBankTransactions($user->account->bank_integration_account_id, $bank_integration))->handle();
-        });
+        if (Ninja::isHosted() && $account->isPaid() && $account->plan == 'enterprise') {
+            $account->bank_integrations()->where('integration_type', BankIntegration::INTEGRATION_TYPE_YODLEE)->where('auto_sync', true)->where('is_deleted', false)->cursor()->each(function ($bank_integration) use ($account) {
```
@turbo124 I added the where check is_deleted. should it even be required in all cases?

paulwer committed:
minor cleanup

paulwer committed:
cleanup

paulwer commented on app/Jobs/Ninja/BankTransactionSync.php:
```diff
@@ -47,17 +49,35 @@ public function handle()
         foreach (MultiDB::$dbs as $db) {
```
@turbo124 do we need any changes here for selfhosted non-multidb envs

turbo124 replied:
The way I have managed this so far is to have two loops:

        if (! config('ninja.db.multi_db_enabled')) {
//self hosted
return
}

//hosted multidb loops

paulwer committed:
reactivate softDeleted Accounts & revoke where deleted_at query
- softDelete, when reconnecting to avoid double transactions
- remove deleted_at query because unnecessary

paulwer commented:
@turbo124 I have requested another review. 
From my pov there are just the open comments, api-documentation and the issue addressed via chat (currency display on account-details) open.

Do you have anything else to add? Have you checked everything from my list above: https://github.com/invoiceninja/invoiceninja/pull/9004#issuecomment-1854262820

turbo124 commented:
@paulwer 

when deleting bank integrations, you'll want to use the methods within BankIntegrationRepository() for archive / restore / delete

turbo124 commented on app/Http/Controllers/Bank/NordigenController.php:
```diff
@@ -0,0 +1,304 @@
+        // prevent rerun of this method with same ref
+        Cache::delete($data["ref"]);
+
+        // Successfull Response => Redirect
```
Use this here:

        return isset($context["is_react"]) && $context['is_react'] ? redirect(config('ninja.react_url'). "/#/settings/bank_accounts") :  redirect(config('ninja.app_url'));

paulwer replied:
i changed the the implementation at the RequestLevel for default behavior, when this prop is not provided, to match default redirect in the react/default application

turbo124 commented on app/Helpers/Bank/Nordigen/Nordigen.php:
```diff
@@ -0,0 +1,118 @@
+            return $it->transform($out);
+        } catch (\Exception $e) {
+            if (strpos($e->getMessage(), "Invalid Account ID") !== false)
+                return null;
```
In Yodlee, at this point we retrun false

turbo124 commented on app/Helpers/Bank/Nordigen/Transformer/TransactionTransformer.php:
```diff
@@ -0,0 +1,137 @@
+      'transaction_id' => $transaction["transactionId"],
+      'amount' => abs((int) $transaction["transactionAmount"]["amount"]),
+      'currency_id' => $this->convertCurrency($transaction["transactionAmount"]["currency"]),
+      'category_id' => 0, // TODO: institution specific keys like: GUTSCHRIFT, ABSCHLUSS, MONATSABSCHLUSS etc
```
we should use null here for category_id if nothing is being inserted

turbo124 commented on app/Http/Requests/Nordigen/ConnectNordigenBankIntegrationRequest.php:
```diff
@@ -0,0 +1,84 @@
+        return [
+            'lang' => 'string',
+            'institution_id' => 'string',
+            'redirect' => 'string', // TODO: @turbo124 @todo validate, that this is a url without / at the end
```
If this is not user input (which for one_time_token it never is) then you can ignore validation here as it has been prebuilt in the PrepareForValidation method

paulwer replied:
the flow can also be started with external one_time_tokens generated by postman for example or external clients. for this case the enduser can provide a redirect.
Also when a specific account would be refreshed after it gets ivalid, the frontend also might provide a different redirect url

turbo124 replied:
In production this would be protected by the api token, and also these fields would be passed in via the nordigen response, so I don't see much value here to use validation.

paulwer replied:
I guess we could remove the value completly, but I had placed it there to be used, when the frontend needs a different redirect. Like I said f.ex. for refreshing an account and redirect to the same account afterwards or be able to check if it was successful for this account.
Or do you mean just to validate if its an url? If its wrong it wouldnt matter?

turbo124 commented on app/Http/Requests/Nordigen/ConnectNordigenBankIntegrationRequest.php:
```diff
@@ -0,0 +1,84 @@
+    {
+        $input = $this->all();
+
+        if (!array_key_exists('redirect', $input)) {
```
As above, the one time token is our security layer here.

paulwer replied:
i did not meant this property for security, it was just how we can validate the provided input.
I will close this conversation, because we can discuss this on the others

turbo124 commented on app/Jobs/Bank/MatchBankTransactions.php:
```diff
@@ -81,6 +81,7 @@ public function __construct(int $company_id, string $db, array $input)
      * Execute the job.
      *
      */
+    // TODO: what are these categories, and for what do we need them
```
These are Yodlee Default Bank Categories, we check here is there is a cached set for reuse.

Now that we have a new provider, we'll need to use additional logic to see whether the appropriate set of categories are present.
 For Nordigen you'll want to do a

paulwer replied:
if i read correctly, nordigen bank account api does not have categories., they have additional apis for this i guess to remember

turbo124 replied:
ok, lets just ensure we set null for the category prop.

paulwer replied:
yes, we do
![image](https://github.com/invoiceninja/invoiceninja/assets/52678724/b99780d3-d5d4-401e-ba7f-3def0bd9b347)

paulwer replied:
you may look at the category_type prop, if we should fill it or not?!
https://developer.gocardless.com/bank-account-data/transactions

paulwer replied:
field-info: Might be used by the financial institution to transport additional transaction related information

turbo124 replied:
I don't think we have anywhere to store this information currently. Perhaps we can look into this in a second pass if needed?

paulwer replied:
sounds great. I close this threat.

turbo124 commented on app/Jobs/Bank/ProcessBankTransactionsNordigen.php:
```diff
@@ -0,0 +1,183 @@
+        try {
+            $this->processTransactions();
+        } catch (\Exception $e) {
+            // reset from_date in case this was the error (self-heal) and perform a max sync of data we can fetch (next-time) @todo we should analyze the error for this
```
I think for Nordigen the most common scenario may be a revoked requisition ? 

There should be checks prior to arriving in this method for that particular scenario (disabled_upstream). So i would assume any failures at this point may cause the disabled_upstream flag to be foggled.

paulwer replied:
the eypired requisition is checked by the isAccountActive Section and we would never reach the Transactions section.

The user can manualy set the fram_date in the portal, so its possible he/she sets it to 3 years ago

paulwer replied:
i reproduced exactly this behavior and nordigen does not return any errors and silently just provides the data within the available time-range.
Should we leave the reset there or remove it?

turbo124 replied:
If it doesn't do anything useful, we can remove it.

paulwer replied:
fixxed

turbo124 commented on app/Jobs/Bank/ProcessBankTransactionsNordigen.php:
```diff
@@ -0,0 +1,183 @@
+            $this->bank_integration->save();
+            $this->stop_loop = false;
+            nlog("Nordigen: account inactive: " . $this->bank_integration->nordigen_account_id);
+            // @turbo124 @todo send email for expired account
```
I'll add this to my todo, do you know if there is a standard block of text that nordigen has for this? Or do we need to create this content ourselves?

paulwer replied:
I did not found any resources, that nordigen handles such cases, we have to send it ourselfes I guess. Maybe we should also add a field in the bank_integration, giving the frontend information, how long the integration will probably last, so the frontend can display a hint to the user to refresh connection a few days before.

Requisitions contain access for 90 days by default.

turbo124 commented on routes/api.php:
```diff
@@ -399,7 +400,9 @@
-    Route::post('api/v1/yodlee/status/{account_number}', [YodleeController::class, 'accountStatus']);
+    Route::post('api/v1/yodlee/status/{account_number}', [YodleeController::class, 'accountStatus']); // @todo @turbo124 check route-path?!
+
+    Route::get('nordigen/institutions', [NordigenController::class, 'institutions'])->name('nordigen.institutions');
```
Good pickup, the /api/v1/ is redundant.

paulwer replied:
i fixed it with the latest commit

turbo124 commented on tests/Feature/Bank/YodleeApiTest.php:
```diff
@@ -166,7 +166,7 @@ public function testCategoryPropertyExists()
     //         {
 
-    //             if(!BankIntegration::where('bank_account_id', $account['id'])->where('company_id', $this->company->id)->exists())
+    //             if(!BankIntegration::where('bank_account_id', $account['id'])->where('company_id', $this->company->id)->exists()) // TODO: maybe filter to only yodlee-integrations
```
safe to ignore.

turbo124 reviewed (changes requested):
Minor changes needed

paulwer committed:
fix https://github.com/invoiceninja/invoiceninja/pull/9004#discussion_r1432302185

paulwer committed:
fix https://github.com/invoiceninja/invoiceninja/pull/9004#discussion_r1432302929

paulwer committed:
ignore comment on testing

paulwer committed:
fix https://github.com/invoiceninja/invoiceninja/pull/9004#discussion_r1432308641

paulwer committed:
cleanups

paulwer committed:
language is now persistent over whole flow + fixes

paulwer committed:
Merge remote-tracking branch 'upstream/v5-develop' into feature-nordigen-payment-provider

paulwer committed:
remove auto_sync condition, because manual sync should also sync bank_integrations, which are not marked as auto_sync + better handling for yodlee

paulwer committed:
adding additional fields to be parsed for transaction: description & participator
https://developer.gocardless.com/bank-account-data/transactions

paulwer commented on app/Helpers/Bank/Nordigen/Transformer/TransactionTransformer.php:
```diff
@@ -0,0 +1,149 @@
+        $description = '';
+        if (array_key_exists('remittanceInformationStructured', $transaction))
+            $description = $transaction["remittanceInformationStructured"];
+        else if (array_key_exists('remittanceInformationStructuredArray', $transaction))
+            $description = implode(' \r\n', $transaction["remittanceInformationStructuredArray"]);
+        else if (array_key_exists('remittanceInformationUnstructured', $transaction))
+            $description = $transaction["remittanceInformationUnstructured"];
+        else if (array_key_exists('remittanceInformationUnstructuredArray', $transaction))
+            $description = implode(' \r\n', $transaction["remittanceInformationUnstructuredArray"]);
+        else
+            Log::warning("Missing description for the following transaction: " . json_encode($transaction));
```
@turbo124 when parsing descriptions I save line-breaks with "\r\n". the frontends does not support this content. Should I change something? "< br >" is also not supported by frontends (I also tried this).

turbo124 replied:
@paulwer 

I think \n should be valid markup for new lines.

Are you not seeing this correctly in the react or flutter apps?

paulwer replied:
only tested flutter and \r\n does not do anything. give me some min, i will test only with \n

paulwer replied:
tested it only in flutter and it is also not working with \n.
I migrated programmcode to use \n only for now

### Table: Description with \r\n
![image](https://github.com/invoiceninja/invoiceninja/assets/52678724/911fd1ba-0d5b-42f3-84e2-586b270dd4d9)

### Table: Description with \n
![image](https://github.com/invoiceninja/invoiceninja/assets/52678724/fa0d80cb-1744-4741-a501-85d2642b1b03)

### Edit Mode
![image](https://github.com/invoiceninja/invoiceninja/assets/52678724/ac17d9dc-b86d-4c7f-a14e-f540b4e6a7fc)

### Datebase
![image](https://github.com/invoiceninja/invoiceninja/assets/52678724/e48b85b7-79eb-4ce7-b55d-5229fe134fe4)

paulwer commented:
INFO: with https://github.com/invoiceninja/invoiceninja/pull/9004/commits/88ff7983c806ad8062b31d6464fb0ed1805d04f2 i also implemented some additional parsings from the response of nordigen and saved the creditor/debitor both into the participant fields.

paulwer committed:
fix https://github.com/invoiceninja/invoiceninja/pull/9004#discussion_r1434778938

paulwer committed:
changes related to https://github.com/invoiceninja/invoiceninja/pull/9004#discussion_r1434778295

turbo124 commented on app/Http/Controllers/BankIntegrationController.php:
```diff
@@ -288,12 +328,20 @@ public function removeAccount(AdminBankIntegrationRequest $request, $acc_id)
-        /** @var \App\Models\User $user */
-        $user = auth()->user();
+        /** @var \App\Models\Account $account */
+        $account = auth()->user()->account();
```
this should change to

$account = auth()->user()->account;

paulwer replied:
done

turbo124 reviewed (commented):
Looks good! Minimal changes now.

paulwer committed:
revert yodlee props: https://github.com/invoiceninja/invoiceninja/pull/9004#discussion_r1435703438

paulwer committed:
changes related to https://github.com/invoiceninja/invoiceninja/pull/9004#discussion_r1435704639

paulwer committed:
cleanups + https://github.com/invoiceninja/invoiceninja/pull/9004#discussion_r1435705476

paulwer commented:
> @paulwer
> 
> when deleting bank integrations, you'll want to use the methods within BankIntegrationRepository() for archive / restore / delete

is this still relevant? I dont think we delete/archive etc with another way, other than default over api? (BankIntegrationController)

paulwer commented:
do we need any other changes related to MultiDB environments, vs single environments?

candidates to check from my pov: MatchBankTransactions, BankMatchingService, BankTransactionSync (just reworked from me, with on eof the latest commits)

paulwer committed:
cleanups

turbo124 commented:
Thank you!

shrippen commented:
The recent versions brought along the corresponding UI button to connect but all I get is a popup that says missing credentials. How to use this feature?

paulwer commented:
privide in self-hosted version the env-vars: NORDIGEN_SECRET_ID & NORDIGEN _SECRET_KEY
and restart the container/instance.

You get these credentials from gocardless/nordigen bank account data api developer dashboard.

turbo124 commented:
.env vars requried as per @paulwer 

```
NORDIGEN_SECRET_ID=
NORDIGEN_SECRET_KEY=
NORDIGEN_TEST_MODE=true
```

shrippen commented:
Note: if you set TEST_MODE=true you will only get access to the sandbox account data
